src/encoding/glyph_cache.rs

```diff
@@ -0,0 +1,87 @@
+// Copyright 2022 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     https://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//
+// Also licensed under MIT license, at your choice.
+
+use std::collections::HashMap;
+
+use super::{Encoding, StreamOffsets};
+use crate::glyph::GlyphProvider;
+
+use peniko::{Fill, Style};
+
+#[derive(Copy, Clone, PartialEq, Eq, Hash, Default, Debug)]
+pub struct GlyphKey {
+    pub font_id: u64,
+    pub font_index: u32,
+    pub glyph_id: u32,
+    pub font_size: u32,
+    pub hint: bool,
+}
+
+#[derive(Default)]
+pub struct GlyphCache {
+    pub encoding: Encoding,
+    glyphs: HashMap<GlyphKey, CachedRange>,
+}
+
+impl GlyphCache {
+    pub fn clear(&mut self) {
+        self.encoding.reset(true);
+        self.glyphs.clear();
+    }
+
+    pub fn get_or_insert(
+        &mut self,
+        key: GlyphKey,
+        style: &Style,
+        scaler: &mut GlyphProvider,
+    ) -> Option<CachedRange> {
+        // For now, only cache non-zero filled glyphs so we don't need to keep style
+        // as part of the key.
+        let is_nz_fill = matches!(style, Style::Fill(Fill::NonZero));
+        if is_nz_fill {
+            if let Some(range) = self.glyphs.get(&key) {
```

Comment:
Updated to use the entry API, but I expect this method and the caching strategy will see some changes soon. I'd like to do a better job on this all around, but since this state is currently just dropped at the end of render, for now I focused on doing the minimum to cache encoded glyphs per frame.